src/resource/resource.c

```diff
@@ -451,15 +459,30 @@ int sid_resource_create_io_event_source(sid_resource_t *res, sid_resource_event_
 	return r;
 }
 
-static int _sd_signal_event_handler(sd_event_source *sd_es, const struct signalfd_siginfo *si, void *data)
+static int _sd_signal_event_handler(sd_event_source *sd_es, int sfd, uint32_t revents, void *data)
 {
 	sid_resource_event_source_t *es = data;
-	return ((sid_resource_signal_event_handler_t) es->handler)(es, si, es->data);
+	struct signalfd_siginfo si;
+
+	ssize_t res;
+
+	res = read (sfd, &si, sizeof(si));
+
+	if (res < 0) {
+		log_error("SIGNAL", "failed to read signal");
+		return 1;
+	}
+	if (res != sizeof(si)) {
+		log_error("SIGNAL", "failed to read size of return data");
+		return 1;
+	}
+
+	return ((sid_resource_signal_event_handler_t) es->handler)(es, &si, es->res);
 }
 
-int sid_resource_create_signal_event_source(sid_resource_t *res, sid_resource_event_source_t **es, int signal,
+int sid_resource_create_signal_event_source(sid_resource_t *res, sid_resource_event_source_t **es, sigset_t mask,
                                             sid_resource_signal_event_handler_t handler, int64_t prio,
-                                            const char *name, void *data)
+                                            const char *name)
```

Comment:
The custom arg passed to the handler - the `void *data` got lost here?

Reply:
I'll fix that.